keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/NotificationEventListener.kt

```diff
@@ -56,30 +67,156 @@ class NotificationEventListener(
+
+      config.environments.forEach { environment ->
+        environment.notifications.filter {
+          it.type == NotificationType.slack
+        }.onEach {
+          handler.sendMessage(SlackMarkAsBadNotification(
+            vetoedArtifact = vetoedArtifact,
+            user = user,
+            targetEnvironment = veto.targetEnvironment,
+            channel = it.address,
+            time = clock.instant(),
+            application = config.name,
+            comment = veto.comment
+          ))
+        }
+      }
+    }
+  }
+
+  @EventListener(ApplicationActuationPaused::class)
+  fun onApplicationActuationPaused(notification: ApplicationActuationPaused) {
+    with(notification) {
+      val config = repository.getDeliveryConfigForApplication(application)
+
+      val handler = handlers.find {
+        it.type == Type.PAUSED_APPLICATION
+      } as? SlackNotificationHandler<SlackPausedNotification> ?: return
+
+      config.environments.forEach { environment ->
+        environment.notifications.filter {
+          it.type == NotificationType.slack
+        }.onEach {
+          handler.sendMessage(SlackPausedNotification(
+            user = triggeredBy,
+            channel = it.address,
+            time = clock.instant(),
+            application = application
+          ))
+        }
+      }
+    }
+  }
+
+  @EventListener(ApplicationActuationResumed::class)
+  fun onApplicationActuationResumed(notification: ApplicationActuationResumed) {
+    with(notification) {
+      val config = repository.getDeliveryConfigForApplication(application)
+
+      val handler = handlers.find {
+        it.type == Type.PAUSED_APPLICATION
+      } as? SlackNotificationHandler<SlackResumedNotification> ?: return
+
+      config.environments.forEach { environment ->
+        environment.notifications.filter {
+          it.type == NotificationType.slack
+        }.onEach {
+          handler.sendMessage(SlackResumedNotification(
+            user = triggeredBy,
+            channel = it.address,
+            time = clock.instant(),
+            application = application
+          ))
+        }
+      }
+    }
+  }
+
+  @EventListener(LifecycleEvent::class)
+  fun onLifecycleEvent(notification: LifecycleEvent) {
+    with(notification) {
+      val config = repository.getDeliveryConfig(artifactRef.split(":")[0])
+      val deliveryArtifact = config.artifacts.find {
+        it.reference == artifactRef.split(":")[1]
+      } ?: return
+
+      val artifact = repository.getArtifactVersion(deliveryArtifact, artifactVersion, null)
+      if (artifact == null) {
+        log.warn("vetoedArtifact is null for application ${config.application}. Can't send MarkAsBadNotification notification")
+        return
+      }
+
+      val handler = handlers.find {
+        it.type == Type.LIFECYCLE_EVENT
+      } as? SlackNotificationHandler<SlackLifecycleNotification>? ?: return
+
+      config.environments.forEach { environment ->
```

Comment:
Same comment/suggestion on looping over environments.

Reply:
same for here - we should notify all environments

Reply:
Hmm, I would think that the lifecycle events should only notify for the environments in which the corresponding artifact is used. I believe there are some utility functions to find out which artifact belongs to which environment.

Reply:
fixed!